### PR TITLE
feat: add backend support for unpublished source datasets (#131)

### PR DESCRIPTION
**Pull request by hunterckx**
Notes:
- The schema and how it fits in with the form might need some attention
  - Currently (with the new fields and no update to the form) the save button for some reason doesn't do anything
- API response datasets are flat objects with several properties used for both published and unpublished source datasets
- `firstAuthorPrimaryName` is renamed to `referenceAuthor` (does this seem appropriate?)
- Field names for new unpublished source dataset are ~~`author`~~ `referenceAuthor`, `contactEmail`, and `title`
  - Would `authorEmail` be more accurate? (I'm inclined not to use just `email` because I think it should be more specific on the API response and ideally that would be consistent with the request)
- ~~The only API update is the `create` API (do we need an API for updating source datasets for this ticket?)~~
  - There is now an edit API

EDIT: the API schema accepts either an object of the form `{ doi: string }`, or of the form `{ contactEmail: string; referenceAuthor: string; title: string }` (~~though I guess we may change `author` to `referenceAuthor`?~~ done!). If DOI is a string, the other three properties must be `undefined`/omitted, and if DOI is `undefined`/omitted, the other three properties must be strings.

### CODE REVIEW

**Comment by NoopDog** [comment]
@hunterckx 

1. can you document the schema for the POST?
2. Can you implement an update API (PUT) for this?
3. Looks like this PR needs a rebase

File: app/services/database.ts
Comment (hunterckx) [inline app/services/database.ts:1]: Basic database interactions are moved to their own file, which causes some changes in otherwise unrelated files

File: app/utils/api-handler.ts [hunk 1]
```diff
@@ -17,22 +14,32 @@ export type MiddlewareFunction = (
...
-      await f(req, res, () => (done = false));
-      if (done) return;
+    try {
+      for (const f of funcs) {
+        let done = true;
+        await f(req, res, () => (done = false));
+        if (done) return;
+      }
+    } catch (e) {
+      respondError(res, e);
```
Comment (hunterckx) [inline app/utils/api-handler.ts:42]: Handlers created by `handler` and `handleByMethod` now automatically handle errors, allowing the functions in `app/services/source-datasets.ts` to throw appropriate errors and have them translate to responses from the API

File: app/views/EditSourceDatasetView/common/schema.ts
Comment (hunterckx) [inline app/views/EditSourceDatasetView/common/schema.ts:1]: I've copied the old versions of the source dataset schemas to be used by the forms, so that the forms still work with the changes to the API schemas

File: app/services/source-datasets.ts
Comment (hunterckx) [inline app/services/source-datasets.ts:1]: `app/services/source-datasets.ts` contains code shared between source dataset creation and editing

File: pages/api/atlases/[atlasId]/source-datasets/[sdId].ts [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+import {
+  AccessError,
+  getUserRoleFromAuthorization,
+  handleByMethod,
+  handler,
+  NotFoundError,
+  role,
+} from "../../../../../app/utils/api-handler";
+
+async function getHandler(
```
Comment (hunterckx) [inline pages/api/atlases/[atlasId]/source-datasets/[sdId].ts:23]: The new API route, `/api/atlases/[atlasId]/source-datasets/[sdId]`, has a GET handler as well

**Review by NoopDog — APPROVED** [review]
Epic! Thanks @hunterckx LGTM!
